File: sorento_crm_frontend/app/(protected)/scm/reorder/lib/trajectory.ts
/**
 * Is this demand going to last? The wording layer.
 *
 * > "Just give me a judgment that whether this demand is going to sustain or is going to
 * >  die off. So I can decide should I order more or should I order just enough or should
 * >  I order less."
 *
 * Mirrors `app/services/scm/trajectory.py`, which decides the verdicts; this file only
 * turns them into words. Every sentence is a fact about OUR OWN order book - like the
 * price advice, nothing here claims to see the market.
 */

export type TrajectoryVerdict = 'rising' | 'holding' | 'falling' | 'quiet' | 'no_history';

export interface TrajectoryEntry {
  verdict: TrajectoryVerdict;
  recent_qty: number;
  previous_qty: number;
  change_pct: number | null;
  year_ago_qty: number | null;
  year_change_pct: number | null;
  window_months: number;
  months: { month: string; qty: number }[];
  customers: { customer_name: string; qty: number }[];
  agents: { name: string; qty: number }[];
  agents_available: boolean;
}

export interface TrajectoryPayload {
  windows: { retail_months: number; project_months: number };
  movement_threshold_pct: number;
  series: Record<string, TrajectoryEntry>;
}

/**
 * The forecast advisory on a buy: how many MORE or FEWER units the trend argues for.
 *
 * > "based on the forecast, how much more should i buy, and how much less should i buy,
 * >  think we need to have this level of intelligence also" (user markup, 2026-08-11)
 *
 * The engine's quantity covers COMMITTED demand and stays untouched - an earlier build let
 * the forecast inflate committed orders (a 2-unit order became 15.9) and that was ruled a
 * bug. So this is an ADVISORY the buyer applies with one click, never a silent change:
 * the trend's own %-change (recent window vs the one before), applied to the buy quantity.
 * Capped at the buy quantity itself - "consider doubling" is the loudest sane suggestion,
 * and "consider less" can at most take the buy to zero.
 */
export interface TrendAdvice {
  direction: 'more' | 'less';
  /** Whole units, >= 1, <= the buy quantity. */
  delta: number;
  /** The |%| the windows moved, rounded, for the row's sentence. */
  pct: number;
}

export function trendAdvice(
  t: TrajectoryEntry | undefined,
  buyQty: number,
): TrendAdvice | null {
  if (!t || buyQty <= 0) return null;
  // Only the two verdicts that ARGUE for a different quantity. Holding/quiet/no-history
  // say nothing about buying more or less, and an advisory on them would be noise.
  if (t.verdict !== 'rising' && t.verdict !== 'falling') return null;
  if (t.change_pct === null) return null;
  const pct = Math.abs(t.change_pct);
  const raw = (buyQty * pct) / 100;
  // Rising rounds up (an advisory of 0 more is not an advisory); falling rounds down so
  // "consider less" never quietly suggests skipping outright.
  const delta = Math.min(t.verdict === 'rising' ? Math.ceil(raw) : Math.floor(raw), buyQty);
  if (delta < 1) return null;
  return {
    direction: t.verdict === 'rising' ? 'more' : 'less',
    delta,
    pct: Math.round(pct),
  };
}

/** The Trend line on the row: the verdict AND what to do about it, in plain words. */
export const TRAJECTORY_ROW_LABEL: Record<TrajectoryVerdict, string> = {
  rising: 'Orders rising - consider more',
  holding: 'Orders steady',
  falling: 'Orders falling - consider less',
  quiet: 'No recent orders - buy just enough',
  no_history: 'Never ordered before',
};

export const TRAJECTORY_TONE: Record<TrajectoryVerdict, 'ok' | 'neutral' | 'warning'> = {
  rising: 'ok',
  holding: 'neutral',
  falling: 'warning',
  quiet: 'warning',
  no_history: 'neutral',
};

/**
 * The verdict with its evidence, for the popup headline.
 *
 * Numbers carry their meaning: "120 in the last 12 months, 90 in the 12 before" reads;
 * a bare +33% does not (AC-S13d.3: never a bare percentage as the headline).
 */
export function describeTrajectory(t: TrajectoryEntry | undefined): string {
  if (!t) return 'No order information for this product.';
  const w = t.window_months;
  const windowWord = `${w} month${w === 1 ? '' : 's'}`;
  const base = `${fmtQty(t.recent_qty)} ordered in the last ${windowWord}, ${fmtQty(t.previous_qty)} in the ${windowWord} before`;

  switch (t.verdict) {
    case 'rising':
      return `Orders are rising: ${base}. Worth considering more than the bare need.`;
    case 'falling':
      return `Orders are falling: ${base}. Consider buying less than the suggestion.`;
    case 'quiet':
      return `Orders stopped: nothing in the last ${windowWord}, after ${fmtQty(t.previous_qty)} in the ${windowWord} before. Buy just enough.`;
    case 'no_history':
      return `No orders on record in the last ${w * 2} months.`;
    case 'holding':
    default:
      return `Orders are steady: ${base}.`;
  }
}

/** The year-ago comparison, beside the verdict - named absent when the book is younger. */
export function describeYearAgo(t: TrajectoryEntry | undefined): string | null {
  if (!t) return null;
  if (t.year_ago_qty === null) return 'No figure for the same window last year.';
  if (t.year_change_pct === null) {
    return `Same window last year: ${fmtQty(t.year_ago_qty)}.`;
  }
  const dir = t.year_change_pct >= 0 ? 'up' : 'down';
  return `Against the same window last year (${fmtQty(t.year_ago_qty)}): ${dir} ${Math.abs(t.year_change_pct).toFixed(0)}%.`;
}

function fmtQty(n: number): string {
  return n.toLocaleString(undefined, { maximumFractionDigits: 0 });
}

/** The key both the plan row and the trajectory map agree on. */
export function trajectoryKey(
  productId: string | null,
  segment: string | null | undefined,
): string | null {
  if (!productId) return null;
  // The service defaults an unsegmented warehouse to 'project', and so must the reader.
  return `${productId}:${segment || 'project'}`;
}

/**
 * The two chart series: this year's months, and the same months a year earlier.
 *
 * > "i need this kind of trend to be in a line graph so it is easier to relate"
 *
 * Both series share the recent 12 months' labels; the year-ago line is the SAME month one
 * year back, so the eye compares seasons vertically.
 */
export function chartSeries(t: TrajectoryEntry): {
  labels: string[];
  thisYear: number[];
  lastYear: (number | null)[];
} {
  const byMonth = new Map(t.months.map((m) => [m.month, m.qty]));
  const labels: string[] = [];
  const now = new Date();
  const months: { y: number; m: number }[] = [];
  for (let i = 11; i >= 0; i--) {
    const d = new Date(now.getFullYear(), now.getMonth() - 1 - i, 1);
    months.push({ y: d.getFullYear(), m: d.getMonth() + 1 });
  }
  const key = (y: number, m: number) => `${y.toString().padStart(4, '0')}-${m.toString().padStart(2, '0')}`;
  const thisYear = months.map(({ y, m }) => byMonth.get(key(y, m)) ?? 0);
  const lastYear = months.map(({ y, m }) => {
    const k = key(y - 1, m);
    return byMonth.has(k) ? (byMonth.get(k) as number) : null;
  });
  for (const { y, m } of months) {
    labels.push(`${String(m).padStart(2, '0')}/${String(y).slice(2)}`);
  }
  return { labels, thisYear, lastYear };
}

File: sorento_crm_backend/app/services/scm/trajectory.py
"""Is this demand sustaining or dying off? The pure arithmetic.

> "Just give me a judgment that whether this demand is going to sustain or is going to die
>  off. So I can decide should I order more or should I order just enough or should I
>  order less."

The judgment is a comparison of our own order history, two ways at once (user decision
2026-08-10): the recent window against the window immediately before it - the comparison
that LEADS - and against the same window last year, which sits beside it and is named
absent when the history is shorter than a year.

Pure by design, like `coverage_timeline` and `cover_service`: the resolver half
(`trajectory_service`) reads the tables and feeds this, so the verdicts stay golden-
testable with no database in the picture.
"""
from __future__ import annotations

from typing import NamedTuple, Optional

#: Below this, two windows differ by noise, not by a trend worth telling a buyer about.
#: Returned in every payload so the screen states the rule it applied.
MOVEMENT_THRESHOLD_PCT = 20.0

#: Default windows when no policy configures them. Retail is stable, so a quarter reads it;
#: project is erratic, so it takes a year to mean anything. Both live on
#: `scm.reorder_policy` and these are only the values for a policy that has not said.
DEFAULT_RETAIL_MONTHS = 3
DEFAULT_PROJECT_MONTHS = 12


class Trajectory(NamedTuple):
    """The verdict and the two comparisons behind it."""

    #: rising | holding | falling | quiet | no_history
    verdict: str
    recent_qty: float
    previous_qty: float
    #: Percent change vs the window immediately before. None when the previous window is
    #: zero - a percentage off nothing is undefined, though the direction is still real.
    change_pct: Optional[float]
    #: Percent change vs the same window last year. None when the book is younger.
    year_ago_qty: Optional[float]
    year_change_pct: Optional[float]
    movement_threshold_pct: float = MOVEMENT_THRESHOLD_PCT


def _pct(new: float, old: float) -> Optional[float]:
    if old == 0:
        return None
    return round((new - old) / abs(old) * 100.0, 2)


def assess_trajectory(
    *, recent: float, previous: float, year_ago: Optional[float]
) -> Trajectory:
    """Turn two (or three) window totals into a verdict.

    Order of the rules:

    * nothing anywhere -> ``no_history``. Not "quiet": quiet means it STOPPED, and a
      product that never sold has no heyday to have stopped from.
    * activity before, nothing now -> ``quiet``. The strongest die-off signal there is.
    * a move past the threshold -> ``rising`` / ``falling``.
    * otherwise ``holding``.
    """
    change = _pct(recent, previous)
    year_change = (
        _pct(recent, year_ago) if year_ago is not None and year_ago > 0
        else None
    )

    if recent == 0 and previous == 0:
        verdict = "no_history"
    elif recent == 0:
        verdict = "quiet"
    elif previous == 0:
        # Started from nothing: direction is real even though the percentage is not.
        verdict = "rising"
    elif change is not None and change >= MOVEMENT_THRESHOLD_PCT:
        verdict = "rising"
    elif change is not None and change <= -MOVEMENT_THRESHOLD_PCT:
        verdict = "falling"
    else:
        verdict = "holding"

    return Trajectory(
        verdict=verdict,
        recent_qty=recent,
        previous_qty=previous,
        change_pct=change,
        year_ago_qty=year_ago,
        year_change_pct=year_change,
    )
